cocotb/binary.py

```diff
@@ -325,36 +330,41 @@ def is_resolvable(self):
         return not any(char in self._str for char in BinaryValue._resolve_to_error)
 
     @property
-    def buff(self):
-        """Attribute :attr:`buff` represents the value as a binary string buffer.
+    def buff(self) -> bytes:
+        r"""The value as a binary string buffer.
 
-        >>> "0100000100101111".buff == "\x41\x2F"
+        >>> BinaryValue("01000001" + "00101111").buff == b"\x41\x2F"
         True
+
+        .. versionchanged:: 1.4
+            This changed from :class:`str` to :class:`bytes`.
+            Note that for older versions used with Python 2 these types were
+            indistinguishable.
         """
         bits = resolve(self._str)
 
         if len(bits) % 8:
             bits = "0" * (8 - len(bits) % 8) + bits
 
-        buff = ""
+        buff = []
         while bits:
             byte = bits[:8]
             bits = bits[8:]
             val = int(byte, 2)
             if self.big_endian:
-                buff += chr(val)
+                buff += [val]
             else:
-                buff = chr(val) + buff
-        return buff
+                buff = [val] + buff
+        return bytes(buff)
 
     @buff.setter
-    def buff(self, val):
+    def buff(self, val: bytes):
         self._str = ""
         for char in val:
             if self.big_endian:
-                self._str += "{0:08b}".format(ord(char))
+                self._str += "{0:08b}".format(char)
             else:
-                self._str = "{0:08b}".format(ord(char)) + self._str
+                self._str = "{0:08b}".format(char) + self._str
```

Comment:
Alternatively,

```python
val = int.from_bytes(val, byteorder=('big' if self.big_endian else 'little'))
self._str = format(val, "b")
```

Reply:
I don't disagree, I just don't want to mix behavior changes and cleanups in the same commit, because it makes them hard to tell apart. The simplification is nice, but I'd want to write some tests to verify the behavior is unchanged.

Reply:
Great, thanks!

Reply:
Well, maybe I do disagree - it seems a bit silly to convert from bytes to bitstring via an arbitrary-precision integer. Will write the test anyway.